File: tests/python/test_schedulers.py
from __future__ import annotations

import math

import optlib


def test_learning_rate_step_schedule() -> None:
    assert (
        optlib.LearningRateAt(0, "step", initial_learning_rate=0.1, gamma=0.5, step_size=3) == 0.1
    )
    assert (
        optlib.LearningRateAt(3, "step", initial_learning_rate=0.1, gamma=0.5, step_size=3) == 0.05
    )


def test_learning_rate_exponential_schedule() -> None:
    value = optlib.LearningRateAt(
        5,
        "exponential",
        initial_learning_rate=0.2,
        decay_rate=0.1,
    )
    assert math.isclose(value, 0.2 * math.exp(-0.5))


def test_learning_rate_cosine_warmup() -> None:
    value = optlib.LearningRateAt(
        0,
        "cosine",
        initial_learning_rate=0.1,
        minimum_learning_rate=0.01,
        total_iterations=10,
        warmup_steps=2,
    )
    assert math.isclose(value, 0.05)

File: tests/cpp/TestSchedulers.cpp
#include <gtest/gtest.h>

#include <optlib/core/schedulers/LearningRate.hpp>

#include <cmath>

TEST(SchedulerTest, ComputesStepSchedule)
{
    optlib::LearningRateConfig config;
    config.Schedule = optlib::LearningRateSchedule::Step;
    config.InitialLearningRate = 0.1;
    config.Gamma = 0.5;
    config.StepSize = 3;

    EXPECT_DOUBLE_EQ(optlib::LearningRateAt(0, config), 0.1);
    EXPECT_DOUBLE_EQ(optlib::LearningRateAt(2, config), 0.1);
    EXPECT_DOUBLE_EQ(optlib::LearningRateAt(3, config), 0.05);
    EXPECT_DOUBLE_EQ(optlib::LearningRateAt(6, config), 0.025);
}

TEST(SchedulerTest, ComputesExponentialSchedule)
{
    optlib::LearningRateConfig config;
    config.Schedule = optlib::LearningRateSchedule::Exponential;
    config.InitialLearningRate = 0.2;
    config.DecayRate = 0.1;

    EXPECT_NEAR(optlib::LearningRateAt(5, config), 0.2 * std::exp(-0.5), 1e-15);
}

TEST(SchedulerTest, ComputesCosineScheduleWithWarmup)
{
    optlib::LearningRateConfig config;
    config.Schedule = optlib::LearningRateSchedule::Cosine;
    config.InitialLearningRate = 0.1;
    config.MinimumLearningRate = 0.01;
    config.TotalIterations = 10;
    config.WarmupSteps = 2;

    EXPECT_NEAR(optlib::LearningRateAt(0, config), 0.05, 1e-15);
    EXPECT_NEAR(optlib::LearningRateAt(10, config), 0.01, 1e-15);
}
